## bowtie2-nf 
a Nextflow pipeline for mapping fastq reads against a reference with [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml) - from building reference index to summarising mapping results.

## How the pipeline works in details
The main workflow can be found in [main.nf](https://github.com/Phuong-Le/bowtie2-nf/blob/main/main.nf)

1. If any of the 6 index files generated by `bowtie2-build` do not exist, then build indices. Otherwise, proceed to mapping. Source code is in [modules/indexReference.nf](https://github.com/Phuong-Le/bowtie2-nf/blob/main/modules/indexReference.nf) 

2. Mapping of reads, this will produce some raw sam files. Source code is in [modules/alnReads.nf](https://github.com/Phuong-Le/bowtie2-nf/blob/main/modules/alnReads.nf) 

3. Processing the reads, including sorting the raw sam files, and summerise them in a file called "aln_summary.txt". Source code is in [modules/processReads.nf](https://github.com/Phuong-Le/bowtie2-nf/blob/main/modules/processReads.nf) 

## Dependencies
- [Nextflow](https://www.nextflow.io/)
- [Docker](https://www.docker.com/) if using own machine or [Singularity](https://sylabs.io/singularity/) if using a shared HPC
- [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml)
- [samtools](https://hcc.unl.edu/docs/applications/app_specific/bioinformatics_tools/data_manipulation_tools/samtools/running_samtools_commands/)


## Installation
- Clone this repo
``` 
git clone git@github.com:Phuong-Le/bowtie2-nf.git
```


## Usage
```
nf_script=/path/to/main.nf
config_file=/path/to/customed_config/file #eg lsf.config
ref_file=/path/to/reference/fasta/file
ref_name=/prefix/of/index/files # ie index name
index_dir=/path/to/dir/containing/index/files
sample_params=/path/to/sample/params/file # a tsv file with 3 columns, sample name, path to fastq1 and path to fastq2
outdir=/path/to/outdir # outdir should contain 2 subdirs with the sam files, and the aln_summary.txt file

"nextflow run ${nf_script} -c ${config_file} \
--ref_file ${ref_file} --ref_name ${ref_name} --index_dir ${index_dir}  \
--sample_params ${sample_params} --outdir ${outdir}"
```

example on an lsf system like at Sanger
```
bsub -cwd /path/to/working_dir -o %J.out -e %J.err -R "select[mem>1000] rusage[mem=1000]" -M1000 \
    "nextflow run ${nf_script} -c ${config_file} \
        --ref_file ${ref_file} --ref_name ${ref_name} --index_dir ${index_dir}  \
        --sample_params ${sample_params} --outdir ${outdir}"
```

demo file for `${sample_params}` is found in [demo_files](https://github.com/Phuong-Le/bowtie2-nf/blob/main/demo_files/sample_params.tsv)


## Authors 
[Phuong Le](https://github.com/Phuong-Le) (email: al35@sanger.ac.uk) 

## Acknowledgements
[Vicky Carr](https://github.com/blue-moon22)

